File: sdks/ruby/lib/rust_pipe.rb
require "json"
require "websocket-client-simple"
require "securerandom"

module RustPipe
  class Worker
    def initialize(url:, handlers:, worker_id: nil, max_concurrency: 10, heartbeat_interval: 5)
      @url = url
      @handlers = handlers
      @worker_id = worker_id || "worker-ruby-#{SecureRandom.hex(4)}"
      @max_concurrency = max_concurrency
      @heartbeat_interval = heartbeat_interval
      @active_tasks = 0
      @start_time = Time.now
      @running = false
      @mutex = Mutex.new
      @semaphore = Queue.new
      max_concurrency.times { @semaphore << true }
    end

    def start
      @running = true
      connect
    end

    def stop
      @running = false
      @ws&.close
    end

    private

    def connect
      worker = self
      @ws = WebSocket::Client::Simple.connect(@url)

      @ws.on :open do
        worker.send(:register)
        worker.send(:start_heartbeat)
      end

      @ws.on :message do |msg|
        worker.send(:handle_message, msg.data)
      end

      @ws.on :close do
        worker.send(:reconnect) if worker.instance_variable_get(:@running)
      end

      sleep # Keep main thread alive
    end

    def register
      send_msg({
        type: "WorkerRegister",
        registration: {
          workerId: @worker_id,
          supportedTasks: @handlers.keys,
          maxConcurrency: @max_concurrency,
          language: "Ruby",
        }
      })
    end

    def start_heartbeat
      Thread.new do
        while @running
          sleep @heartbeat_interval
          send_msg({
            type: "Heartbeat",
            payload: {
              workerId: @worker_id,
              activeTasks: @active_tasks,
              capacity: @max_concurrency,
              uptimeSeconds: (Time.now - @start_time).to_i,
            }
          })
        end
      end
    end

    def handle_message(raw)
      msg = JSON.parse(raw, symbolize_names: true)
      case msg[:type]
      when "TaskDispatch"
        task = msg[:task]
        Thread.new { execute_task(task) }
      when "Shutdown"
        stop
      end
    rescue JSON::ParserError
      nil
    end

    def execute_task(task)
      @semaphore.pop
      @mutex.synchronize { @active_tasks += 1 }
      start_time = Time.now

      handler = @handlers[task[:taskType]]
      unless handler
        send_result(task[:id], "Failed", nil,
          { code: "UNKNOWN_TASK_TYPE", message: "No handler for: #{task[:taskType]}", retryable: false }, 0)
        return
      end

      result = handler.call(task)
      duration = ((Time.now - start_time) * 1000).to_i
      send_result(task[:id], "Completed", result, nil, duration)
    rescue => e
      duration = ((Time.now - start_time) * 1000).to_i
      send_result(task[:id], "Failed", nil,
        { code: "HANDLER_ERROR", message: e.message, retryable: true }, duration)
    ensure
      @mutex.synchronize { @active_tasks -= 1 }
      @semaphore << true
    end

    def send_result(task_id, status, payload, error, duration_ms)
      send_msg({
        type: "TaskResult",
        result: {
          taskId: task_id,
          status: status,
          payload: payload,
          error: error,
          durationMs: duration_ms,
          workerId: @worker_id,
        }
      })
    end

    def send_msg(msg)
      @ws&.send(JSON.generate(msg))
    end
  end
end

File: sdks/elixir/lib/rust_pipe.ex
defmodule RustPipe.Worker do
  use WebSockex
  require Logger

  defstruct [:worker_id, :handlers, :max_concurrency, :active_tasks, :start_time]

  def start_link(url, opts) do
    worker_id = Keyword.get(opts, :worker_id, "worker-elixir-#{:crypto.strong_rand_bytes(4) |> Base.encode16(case: :lower)}")
    handlers = Keyword.fetch!(opts, :handlers)
    max_concurrency = Keyword.get(opts, :max_concurrency, 10)

    state = %__MODULE__{
      worker_id: worker_id,
      handlers: handlers,
      max_concurrency: max_concurrency,
      active_tasks: 0,
      start_time: System.monotonic_time(:second)
    }

    WebSockex.start_link(url, __MODULE__, state, name: __MODULE__)
  end

  @impl true
  def handle_connect(_conn, state) do
    register(state)
    schedule_heartbeat(state)
    {:ok, state}
  end

  @impl true
  def handle_frame({:text, raw}, state) do
    case Jason.decode(raw) do
      {:ok, %{"type" => "TaskDispatch", "task" => task}} ->
        spawn(fn -> execute_task(task, state) end)
        {:ok, state}

      {:ok, %{"type" => "Shutdown"}} ->
        {:close, state}

      _ ->
        {:ok, state}
    end
  end

  @impl true
  def handle_info(:heartbeat, state) do
    payload = %{
      type: "Heartbeat",
      payload: %{
        workerId: state.worker_id,
        activeTasks: state.active_tasks,
        capacity: state.max_concurrency,
        uptimeSeconds: System.monotonic_time(:second) - state.start_time
      }
    }
    frame = {:text, Jason.encode!(payload)}
    schedule_heartbeat(state)
    {:reply, frame, state}
  end

  defp register(state) do
    msg = %{
      type: "WorkerRegister",
      registration: %{
        workerId: state.worker_id,
        supportedTasks: Map.keys(state.handlers),
        maxConcurrency: state.max_concurrency,
        language: "Elixir"
      }
    }
    WebSockex.send_frame(self(), {:text, Jason.encode!(msg)})
  end

  defp schedule_heartbeat(_state) do
    Process.send_after(self(), :heartbeat, 5_000)
  end

  defp execute_task(task, state) do
    task_id = task["id"]
    task_type = task["taskType"]
    start = System.monotonic_time(:millisecond)

    handler = Map.get(state.handlers, task_type)

    result =
      if handler do
        try do
          output = handler.(task["payload"])
          duration = System.monotonic_time(:millisecond) - start
          %{taskId: task_id, status: "Completed", payload: output, durationMs: duration, workerId: state.worker_id}
        rescue
          e ->
            duration = System.monotonic_time(:millisecond) - start
            %{taskId: task_id, status: "Failed",
              error: %{code: "HANDLER_ERROR", message: Exception.message(e), retryable: true},
              durationMs: duration, workerId: state.worker_id}
        end
      else
        %{taskId: task_id, status: "Failed",
          error: %{code: "UNKNOWN_TASK_TYPE", message: "No handler for: #{task_type}", retryable: false},
          durationMs: 0, workerId: state.worker_id}
      end

    msg = %{type: "TaskResult", result: result}
    WebSockex.send_frame(__MODULE__, {:text, Jason.encode!(msg)})
  end
end
